src/github.com/t3chguy/riot-static/riot-static.go

```diff
@@ -0,0 +1,234 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/*
+	@TODO Show more than just textual events (Images etc)
+	@TODO Historical member info
+	@TODO make Peeking more useful (paginate etc)
+	@TODO investigate best practices for the spider to get the most out of it
+	@TODO make cache invalidate self after N requests of M minutes
+*/
+
+package main
+
+import (
+	"fmt"
+	"github.com/gin-gonic/gin"
+	"github.com/matryer/resync"
+	"net/http"
+	"os"
+	"strconv"
+	"sync"
+)
+
+func paginate(page int, size int, length int) (skip int, end int) {
+	if skip = (page - 1) * size; skip > length {
+		skip = length
+	}
+	if end = skip + size; end > length {
+		end = length
+	}
+	return
+}
+
+func GetPublicRoomsList(c *gin.Context) {
+	var page int
+	var err error
+	if page, err = strconv.Atoi(c.DefaultQuery("page", "1")); err != nil {
+		page = 1
+	}
+
+	pageSize := 20
+
+	data.RLock()
+	skip, end := paginate(page, pageSize, data.NumRooms)
+	c.HTML(http.StatusOK, "rooms.html", gin.H{
+		"Rooms":    data.Ordered[skip:end],
+		"NumRooms": data.NumRooms,
+		"Page":     page,
+	})
+	data.RUnlock()
+}
+
+func GetPublicRoom(c *gin.Context) { // ~/room/:roomId/chat
+	roomId := c.Param("roomId")
+
+	data.RLock()
+	c.HTML(http.StatusOK, "room.html", gin.H{
+		"Room": data.Rooms[roomId],
+	})
+	data.RUnlock()
+}
+
+func GetPublicRoomServers(c *gin.Context) { // ~/room/:roomId/servers
+	roomId := c.Param("roomId")
+
+	data.RLock()
+	c.HTML(http.StatusOK, "room_servers.html", gin.H{
+		"Room": data.Rooms[roomId],
+	})
+	data.RUnlock()
+}
+
+func GetPublicRoomMembers(c *gin.Context) { // ~/room/:roomId/members
+	roomId := c.Param("roomId")
+
+	//var page int
+	//var err error
+	//if page, err = strconv.Atoi(c.DefaultQuery("page", "1")); err != nil {
+	//	page = 1
+	//}
+
+	//pageSize := 20
+
+	data.RLock()
+	length := len(data.Rooms[roomId].MemberInfo)
+	//skip, end := paginate(page, pageSize, length)
+	c.HTML(http.StatusOK, "room_members.html", gin.H{
+		"Room": data.Rooms[roomId],
+		//"MemberInfo": data.Rooms[roomId].Members[skip:end],
+		"NumMembers": length,
+	})
+	data.RUnlock()
+}
+
+func GetPublicRoomPowerLevels(c *gin.Context) { // ~/room/:roomId/power_levels
+	roomId := c.Param("roomId")
+
+	data.RLock()
+	c.HTML(http.StatusOK, "power_levels.html", gin.H{
+		"PowerLevels": data.Rooms[roomId].PowerLevels,
+	})
+	data.RUnlock()
+}
+
+func GetPublicRoomMember(c *gin.Context) { // ~/room/:roomId/members/:mxid
+	roomId := c.Param("roomId")
+	mxid := c.Param("mxid")
+
+	data.RLock()
+	if memberInfo := data.Rooms[roomId].MemberInfo[mxid]; memberInfo != nil {
+		c.HTML(http.StatusOK, "member_info.html", gin.H{
+			"RoomID":     roomId,
+			"MXID":       mxid,
+			"MemberInfo": memberInfo,
+		})
+	} else {
+		c.AbortWithStatus(http.StatusNotFound)
+	}
+	data.RUnlock()
+}
+
+var data = struct {
+	resync.Once
+	sync.RWMutex
+	NumRooms int
+	Ordered  []*Room
+	Rooms    map[string]*Room
+}{}
```

Comment:
This is the bulk of your data at runtime. It's absolutely vital you guard it against concurrent access to the `Rooms` map. I would put this in its own file with accessor functions which do this for you, so you don't need to constantly remember when it is safe to lock/unlock.